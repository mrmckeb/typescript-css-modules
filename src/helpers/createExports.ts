import { CSSExports } from 'icss-utils';
import reserved from 'reserved-words';
import { Options } from '../options';
import { transformClasses } from './classTransforms';

const NOT_CAMELCASE_REGEXP = /[\-_]/;

const classNameToProperty = (className: string) => `'${className}': string;`;
const classNameToNamedExport = (className: string) =>
  `export const ${className}: string;`;

const flattenClassNames = (
  previousValue: string[] = [],
  currentValue: string[],
) => previousValue.concat(currentValue);

export const createExports = (classes: CSSExports, options: Options) => {
  const isCamelCase = (className: string) =>
    !NOT_CAMELCASE_REGEXP.test(className);
  const isReservedWord = (className: string) => !reserved.check(className);

  const processedClasses = Object.keys(classes)
    .map(transformClasses(options.classnameTransform))
    .reduce(flattenClassNames, []);
  const camelCasedKeys = processedClasses
    .filter(isCamelCase)
    .filter(isReservedWord)
    .map(classNameToNamedExport);

  const defaultExport = `\
declare const classes: {
${processedClasses.map(classNameToProperty).join('\n  ')}
};
export default classes;
`;

  if (camelCasedKeys.length) {
    return defaultExport + camelCasedKeys.join('\n') + '\n';
  }
  return defaultExport;
};
