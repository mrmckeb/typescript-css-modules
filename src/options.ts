import { Options as SassOptions } from 'sass';
import tsModule from 'typescript/lib/tsserverlibrary';
import { DotenvConfigOptions } from 'dotenv/types';
import { CSSExports } from 'icss-utils';
import { Logger } from './helpers/logger';

import stylus from 'stylus';

type StylusRenderOptions = Parameters<typeof stylus>[1];

export interface PostCssOptions {
  excludePlugins?: string[];
  useConfig?: boolean;
}

export interface RendererOptions {
  less?: Partial<Less.Options>;
  sass?: Partial<SassOptions>;
  stylus?: Partial<StylusRenderOptions>;
}

export interface Options {
  classnameTransform?: ClassnameTransformOptions;
  customMatcher?: string;
  customRenderer?: string;
  customTemplate?: string;
  dotenvOptions?: DotenvConfigOptions;
  postCssOptions?: PostCssOptions;
  rendererOptions?: RendererOptions;
}

export type ClassnameTransformOptions =
  | 'asIs'
  | 'camelCase'
  | 'camelCaseOnly'
  | 'dashes'
  | 'dashesOnly';

export interface CustomRendererOptions {
  fileName: string;
  logger: Logger;
  compilerOptions: tsModule.CompilerOptions;
}

export type CustomRenderer = (
  css: string,
  options: CustomRendererOptions,
) => string;

export interface CustomTemplateOptions {
  classes: CSSExports;
  fileName: string;
  logger: Logger;
}

export type CustomTemplate = (
  dts: string,
  options: CustomTemplateOptions,
) => string;
