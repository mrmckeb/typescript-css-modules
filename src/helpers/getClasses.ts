import path from 'path';
import postcss from 'postcss';
import less from 'less';
import sass from 'sass';
import stylus from 'stylus';
import { extractICSS } from 'icss-utils';
import tsModule from 'typescript/lib/tsserverlibrary';
import { Logger } from './logger';
import { Options, CustomRenderer } from '../options';

export const enum FileTypes {
  css = 'css',
  less = 'less',
  sass = 'sass',
  scss = 'scss',
  styl = 'styl',
}

export const getFileType = (fileName: string) => {
  if (fileName.endsWith('.css')) return FileTypes.css;
  if (fileName.endsWith('.less')) return FileTypes.less;
  if (fileName.endsWith('.sass')) return FileTypes.sass;
  if (fileName.endsWith('.styl')) return FileTypes.styl;
  return FileTypes.scss;
};

const getFilePath = (fileName: string) => path.dirname(fileName);

export const getClasses = ({
  css,
  fileName,
  logger,
  options,
  processor,
  compilerOptions,
}: {
  css: string;
  fileName: string;
  logger: Logger;
  options: Options;
  processor: postcss.Processor;
  compilerOptions: tsModule.CompilerOptions;
}) => {
  try {
    const fileType = getFileType(fileName);
    const rendererOptions = options.rendererOptions || {};

    let transformedCss = '';

    if (options.customRenderer) {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const customRenderer = require(options.customRenderer) as CustomRenderer;
      transformedCss = customRenderer(css, {
        fileName,
        logger,
        compilerOptions,
      });
    } else if (fileType === FileTypes.less) {
      less.render(
        css,
        {
          syncImport: true,
          filename: fileName,
          ...(rendererOptions.less || {}),
        } as Less.Options,
        (error, output) => {
          if (error || output === undefined) throw error;
          transformedCss = output.css.toString();
        },
      );
    } else if (fileType === FileTypes.scss || fileType === FileTypes.sass) {
      const filePath = getFilePath(fileName);
      const { includePaths, ...sassOptions } = rendererOptions.sass || {};

      transformedCss = sass
        .renderSync({
          // NOTE: Solves an issue where tilde imports fail.
          // https://github.com/sass/dart-sass/issues/801
          // Will strip `~` from imports, unless followed by a slash.
          data: css.replace(/(@import ['"])~(?!\/)/gm, '$1'),
          indentedSyntax: fileType === FileTypes.sass,
          includePaths: [filePath, 'node_modules', ...(includePaths || [])],
          ...sassOptions,
        })
        .css.toString();
    } else if (fileType === FileTypes.styl) {
      transformedCss = stylus(css, {
        ...(rendererOptions.stylus || {}),
        filename: fileName,
      }).render();
    } else {
      transformedCss = css;
    }

    const processedCss = processor.process(transformedCss, {
      from: fileName,
    });

    return processedCss.root ? extractICSS(processedCss.root).icssExports : {};
  } catch (e) {
    logger.error(e);
    return {};
  }
};
