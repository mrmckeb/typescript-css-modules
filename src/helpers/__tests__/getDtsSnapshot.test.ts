import { readFileSync } from 'fs';
import { CSSExports } from 'icss-utils';
import { join } from 'path';
import postcss from 'postcss';
import postcssIcssSelectors from 'postcss-icss-selectors';
import postcssImportSync from 'postcss-import-sync2';
import { getClasses } from '../getClasses';
import { createExports } from '../createExports';
import { Logger } from '../logger';
import { Options } from '../../options';

const testFileNames = [
  'test.module.css',
  'test.module.less',
  'test.module.scss',
  'empty.module.less',
  'empty.module.scss',
  'import.module.css',
  'import.module.less',
];

const mockLogger: Logger = {
  log: jest.fn(),
  error: jest.fn(),
};

const mockOptions: Options = {};

const processor = postcss([
  postcssImportSync(),
  postcssIcssSelectors({ mode: 'local' }),
]);

describe('utils / cssSnapshots', () => {
  testFileNames.forEach((fileName) => {
    let classes: CSSExports;
    const fullFileName = join(__dirname, 'fixtures', fileName);
    const testFile = readFileSync(fullFileName, 'utf8');

    beforeAll(() => {
      classes = getClasses(
        processor,
        testFile,
        fullFileName,
        mockOptions,
        mockLogger,
      );
    });

    describe(`with file '${fileName}'`, () => {
      describe('getClasses', () => {
        it('should return an object matching expected CSS', () => {
          expect(classes).toMatchSnapshot();
        });
      });

      describe('createExports', () => {
        it('should create an exports file', () => {
          const exports = createExports(classes, {});
          expect(exports).toMatchSnapshot();
        });
      });
    });
  });

  describe(`with a custom renderer`, () => {
    const fullFileName = 'exampleFileContents';
    const testFile = 'exampleFileName';
    const customRenderer = join(__dirname, 'fixtures', 'customRenderer.js');
    const classes = getClasses(
      processor,
      testFile,
      fullFileName,
      { customRenderer },
      mockLogger,
    );

    it('should process a file and log', () => {
      expect(classes).toMatchSnapshot();
      expect(mockLogger.log).toHaveBeenCalledWith('Example log');
    });
  });

  describe('includePaths in sass options', () => {
    const fullFileName = join(
      __dirname,
      'fixtures',
      'include-path.module.scss',
    );
    const testFile = readFileSync(fullFileName, 'utf8');

    it('should find external file from includePaths', () => {
      const classes = getClasses(
        processor,
        testFile,
        fullFileName,
        {
          rendererOptions: {
            sass: { includePaths: [join(__dirname, 'external')] },
          },
        },
        mockLogger,
      );

      expect(classes).toMatchSnapshot();
    });
  });
});
