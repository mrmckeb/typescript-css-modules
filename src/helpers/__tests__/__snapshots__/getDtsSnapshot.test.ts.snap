// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils / cssSnapshots with a Bootstrap import should find external files 1`] = `"bootstrap-module__test---3KFKN"`;

exports[`utils / cssSnapshots with a custom renderer should process a file and log 1`] = `
Object {
  "exampleFileContents": "exampleFileContents__exampleFileContents---e3Nf2",
  "exampleFileName": "exampleFileContents__exampleFileName---yg7Zl",
}
`;

exports[`utils / cssSnapshots with file 'empty.module.less' createExports should create an exports file 1`] = `
"declare const classes: {

};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'empty.module.less' getClasses should return an object matching expected CSS 1`] = `Object {}`;

exports[`utils / cssSnapshots with file 'empty.module.less' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {

};
export default classes;

export const __cssModule: true;
export type AllClassNames = '';"
`;

exports[`utils / cssSnapshots with file 'empty.module.sass' createExports should create an exports file 1`] = `
"declare const classes: {

};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'empty.module.sass' getClasses should return an object matching expected CSS 1`] = `Object {}`;

exports[`utils / cssSnapshots with file 'empty.module.sass' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {

};
export default classes;

export const __cssModule: true;
export type AllClassNames = '';"
`;

exports[`utils / cssSnapshots with file 'empty.module.scss' createExports should create an exports file 1`] = `
"declare const classes: {

};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'empty.module.scss' getClasses should return an object matching expected CSS 1`] = `Object {}`;

exports[`utils / cssSnapshots with file 'empty.module.scss' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {

};
export default classes;

export const __cssModule: true;
export type AllClassNames = '';"
`;

exports[`utils / cssSnapshots with file 'empty.module.styl' createExports should create an exports file 1`] = `
"declare const classes: {

};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'empty.module.styl' getClasses should return an object matching expected CSS 1`] = `Object {}`;

exports[`utils / cssSnapshots with file 'empty.module.styl' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {

};
export default classes;

export const __cssModule: true;
export type AllClassNames = '';"
`;

exports[`utils / cssSnapshots with file 'import.module.css' createExports should create an exports file 1`] = `
"declare const classes: {
'classA': string;
  'ClassB': string;
  'class-c': string;
  'class_d': string;
  'parent': string;
  'childA': string;
  'childB': string;
  'nestedChild': string;
};
export default classes;
export const classA: string;
export const ClassB: string;
export const parent: string;
export const childA: string;
export const childB: string;
export const nestedChild: string;
"
`;

exports[`utils / cssSnapshots with file 'import.module.css' getClasses should return an object matching expected CSS 1`] = `
Object {
  "ClassB": "import-module__ClassB---1uTmq",
  "childA": "import-module__childA---39Yod",
  "childB": "import-module__childB---2S1Lz",
  "class-c": "import-module__class-c---Xdyrv",
  "classA": "import-module__classA---3-9yK",
  "class_d": "import-module__class_d---27FnC",
  "nestedChild": "import-module__nestedChild---HAsOX",
  "parent": "import-module__parent---11jeJ",
}
`;

exports[`utils / cssSnapshots with file 'import.module.css' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {
'classA': string;
  'ClassB': string;
  'class-c': string;
  'class_d': string;
  'parent': string;
  'childA': string;
  'childB': string;
  'nestedChild': string;
};
export default classes;
export const classA: string;
export const ClassB: string;
export const parent: string;
export const childA: string;
export const childB: string;
export const nestedChild: string;

export const __cssModule: true;
export type AllClassNames = 'classA' | 'ClassB' | 'class-c' | 'class_d' | 'parent' | 'childA' | 'childB' | 'nestedChild';"
`;

exports[`utils / cssSnapshots with file 'import.module.less' createExports should create an exports file 1`] = `
"declare const classes: {
'nested-class-parent': string;
  'child-class': string;
  'selector-blue': string;
  'selector-green': string;
  'selector-red': string;
  'column-1': string;
  'column-2': string;
  'column-3': string;
  'column-4': string;
  'color-set': string;
};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'import.module.less' getClasses should return an object matching expected CSS 1`] = `
Object {
  "child-class": "import-module__child-class---3rUzY",
  "color-set": "import-module__color-set---1Nz_3",
  "column-1": "import-module__column-1---1do6Y",
  "column-2": "import-module__column-2---HjqfU",
  "column-3": "import-module__column-3---N5Wnb",
  "column-4": "import-module__column-4---23ZMh",
  "nested-class-parent": "import-module__nested-class-parent---36rk0",
  "selector-blue": "import-module__selector-blue---1tI9n",
  "selector-green": "import-module__selector-green---2yk8M",
  "selector-red": "import-module__selector-red---1SN6v",
}
`;

exports[`utils / cssSnapshots with file 'import.module.less' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {
'nested-class-parent': string;
  'child-class': string;
  'selector-blue': string;
  'selector-green': string;
  'selector-red': string;
  'column-1': string;
  'column-2': string;
  'column-3': string;
  'column-4': string;
  'color-set': string;
};
export default classes;

export const __cssModule: true;
export type AllClassNames = 'nested-class-parent' | 'child-class' | 'selector-blue' | 'selector-green' | 'selector-red' | 'column-1' | 'column-2' | 'column-3' | 'column-4' | 'color-set';"
`;

exports[`utils / cssSnapshots with file 'import.module.styl' createExports should create an exports file 1`] = `
"declare const classes: {
'foo': string;
  'bar': string;
  'baz': string;
  'col-1': string;
  'col-2': string;
  'col-3': string;
  'local-class-1': string;
  'inside-local': string;
  'inside-1-name-2': string;
  'inside-2-name-1': string;
};
export default classes;
export const foo: string;
export const bar: string;
export const baz: string;
"
`;

exports[`utils / cssSnapshots with file 'import.module.styl' getClasses should return an object matching expected CSS 1`] = `
Object {
  "bar": "import-module__bar---PavMt",
  "baz": "import-module__baz---2h0QO",
  "col-1": "import-module__col-1---k2h_y",
  "col-2": "import-module__col-2---Rvdyk",
  "col-3": "import-module__col-3---bhl8t",
  "foo": "import-module__foo---3m4Gt",
  "inside-1-name-2": "import-module__inside-1-name-2---3GMJE",
  "inside-2-name-1": "import-module__inside-2-name-1---2Q-yK",
  "inside-local": "import-module__inside-local---22ivP",
  "local-class-1": "import-module__local-class-1---3Cqnu",
}
`;

exports[`utils / cssSnapshots with file 'import.module.styl' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {
'foo': string;
  'bar': string;
  'baz': string;
  'col-1': string;
  'col-2': string;
  'col-3': string;
  'local-class-1': string;
  'inside-local': string;
  'inside-1-name-2': string;
  'inside-2-name-1': string;
};
export default classes;
export const foo: string;
export const bar: string;
export const baz: string;

export const __cssModule: true;
export type AllClassNames = 'foo' | 'bar' | 'baz' | 'col-1' | 'col-2' | 'col-3' | 'local-class-1' | 'inside-local' | 'inside-1-name-2' | 'inside-2-name-1';"
`;

exports[`utils / cssSnapshots with file 'test.module.css' createExports should create an exports file 1`] = `
"declare const classes: {
'classA': string;
  'ClassB': string;
  'class-c': string;
  'class_d': string;
  'parent': string;
  'childA': string;
  'childB': string;
  'nestedChild': string;
};
export default classes;
export const classA: string;
export const ClassB: string;
export const parent: string;
export const childA: string;
export const childB: string;
export const nestedChild: string;
"
`;

exports[`utils / cssSnapshots with file 'test.module.css' getClasses should return an object matching expected CSS 1`] = `
Object {
  "ClassB": "test-module__ClassB---1z7xQ",
  "childA": "test-module__childA---3PtDp",
  "childB": "test-module__childB---33c4a",
  "class-c": "test-module__class-c---Nme_S",
  "classA": "test-module__classA---11Lf2",
  "class_d": "test-module__class_d---7rY6T",
  "nestedChild": "test-module__nestedChild---3OSwX",
  "parent": "test-module__parent---1pNsk",
}
`;

exports[`utils / cssSnapshots with file 'test.module.css' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {
'classA': string;
  'ClassB': string;
  'class-c': string;
  'class_d': string;
  'parent': string;
  'childA': string;
  'childB': string;
  'nestedChild': string;
};
export default classes;
export const classA: string;
export const ClassB: string;
export const parent: string;
export const childA: string;
export const childB: string;
export const nestedChild: string;

export const __cssModule: true;
export type AllClassNames = 'classA' | 'ClassB' | 'class-c' | 'class_d' | 'parent' | 'childA' | 'childB' | 'nestedChild';"
`;

exports[`utils / cssSnapshots with file 'test.module.less' createExports should create an exports file 1`] = `
"declare const classes: {
'nested-class-parent': string;
  'child-class': string;
  'selector-blue': string;
  'selector-green': string;
  'selector-red': string;
  'column-1': string;
  'column-2': string;
  'column-3': string;
  'column-4': string;
  'color-set': string;
};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'test.module.less' getClasses should return an object matching expected CSS 1`] = `
Object {
  "child-class": "test-module__child-class---1Ga8C",
  "color-set": "test-module__color-set---2h232",
  "column-1": "test-module__column-1---3Hv-0",
  "column-2": "test-module__column-2---1QOqJ",
  "column-3": "test-module__column-3---1PZz0",
  "column-4": "test-module__column-4---2XpQh",
  "nested-class-parent": "test-module__nested-class-parent---R9jV3",
  "selector-blue": "test-module__selector-blue---1icL9",
  "selector-green": "test-module__selector-green---2pD8c",
  "selector-red": "test-module__selector-red---gsEUj",
}
`;

exports[`utils / cssSnapshots with file 'test.module.less' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {
'nested-class-parent': string;
  'child-class': string;
  'selector-blue': string;
  'selector-green': string;
  'selector-red': string;
  'column-1': string;
  'column-2': string;
  'column-3': string;
  'column-4': string;
  'color-set': string;
};
export default classes;

export const __cssModule: true;
export type AllClassNames = 'nested-class-parent' | 'child-class' | 'selector-blue' | 'selector-green' | 'selector-red' | 'column-1' | 'column-2' | 'column-3' | 'column-4' | 'color-set';"
`;

exports[`utils / cssSnapshots with file 'test.module.sass' createExports should create an exports file 1`] = `
"declare const classes: {
'local-class-inside-global': string;
  'local-class': string;
  'local-class-2': string;
  'local-class-inside-local': string;
  'reserved-words': string;
  'default': string;
  'const': string;
  'nested-class-parent': string;
  'child-class': string;
  'nested-class-parent--extended': string;
  'section-1': string;
  'section-2': string;
  'section-3': string;
  'section-4': string;
  'section-5': string;
  'section-6': string;
  'section-7': string;
  'section-8': string;
  'section-9': string;
  'class-with-mixin': string;
};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'test.module.sass' getClasses should return an object matching expected CSS 1`] = `
Object {
  "child-class": "test-module__child-class---1Z6FR",
  "class-with-mixin": "test-module__class-with-mixin---3VX2U",
  "const": "test-module__const---SwQeI",
  "default": "test-module__default---1XDkC",
  "local-class": "test-module__local-class---3t74b",
  "local-class-2": "test-module__local-class-2---3ZpZR",
  "local-class-inside-global": "test-module__local-class-inside-global---3XSpm",
  "local-class-inside-local": "test-module__local-class-inside-local---2habk",
  "nested-class-parent": "test-module__nested-class-parent---2NZYL",
  "nested-class-parent--extended": "test-module__nested-class-parent--extended---2qy0B",
  "reserved-words": "test-module__reserved-words---75Ick",
  "section-1": "test-module__section-1---SBQkq",
  "section-2": "test-module__section-2---1Z96g",
  "section-3": "test-module__section-3---_1k22",
  "section-4": "test-module__section-4---1xlGf",
  "section-5": "test-module__section-5---19rEJ",
  "section-6": "test-module__section-6---2J7BT",
  "section-7": "test-module__section-7---1oMBc",
  "section-8": "test-module__section-8---3utTC",
  "section-9": "test-module__section-9---t7Agm",
}
`;

exports[`utils / cssSnapshots with file 'test.module.sass' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {
'local-class-inside-global': string;
  'local-class': string;
  'local-class-2': string;
  'local-class-inside-local': string;
  'reserved-words': string;
  'default': string;
  'const': string;
  'nested-class-parent': string;
  'child-class': string;
  'nested-class-parent--extended': string;
  'section-1': string;
  'section-2': string;
  'section-3': string;
  'section-4': string;
  'section-5': string;
  'section-6': string;
  'section-7': string;
  'section-8': string;
  'section-9': string;
  'class-with-mixin': string;
};
export default classes;

export const __cssModule: true;
export type AllClassNames = 'local-class-inside-global' | 'local-class' | 'local-class-2' | 'local-class-inside-local' | 'reserved-words' | 'default' | 'const' | 'nested-class-parent' | 'child-class' | 'nested-class-parent--extended' | 'section-1' | 'section-2' | 'section-3' | 'section-4' | 'section-5' | 'section-6' | 'section-7' | 'section-8' | 'section-9' | 'class-with-mixin';"
`;

exports[`utils / cssSnapshots with file 'test.module.scss' createExports should create an exports file 1`] = `
"declare const classes: {
'local-class-inside-global': string;
  'local-class': string;
  'local-class-2': string;
  'local-class-inside-local': string;
  'reserved-words': string;
  'default': string;
  'const': string;
  'nested-class-parent': string;
  'child-class': string;
  'nested-class-parent--extended': string;
  'section-1': string;
  'section-2': string;
  'section-3': string;
  'section-4': string;
  'section-5': string;
  'section-6': string;
  'section-7': string;
  'section-8': string;
  'section-9': string;
  'class-with-mixin': string;
};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'test.module.scss' getClasses should return an object matching expected CSS 1`] = `
Object {
  "child-class": "test-module__child-class---1X87n",
  "class-with-mixin": "test-module__class-with-mixin---2sqUv",
  "const": "test-module__const---_YsB3",
  "default": "test-module__default---2bBi7",
  "local-class": "test-module__local-class---2mqdW",
  "local-class-2": "test-module__local-class-2---3V2Ud",
  "local-class-inside-global": "test-module__local-class-inside-global---19zZ3",
  "local-class-inside-local": "test-module__local-class-inside-local---2DBWn",
  "nested-class-parent": "test-module__nested-class-parent---38azU",
  "nested-class-parent--extended": "test-module__nested-class-parent--extended---2RArS",
  "reserved-words": "test-module__reserved-words---LxD0z",
  "section-1": "test-module__section-1---h9TrY",
  "section-2": "test-module__section-2---pg5Dq",
  "section-3": "test-module__section-3---141D0",
  "section-4": "test-module__section-4---29Zqe",
  "section-5": "test-module__section-5---2aEf7",
  "section-6": "test-module__section-6---3L42n",
  "section-7": "test-module__section-7---2SoUE",
  "section-8": "test-module__section-8---HwOIP",
  "section-9": "test-module__section-9---3MO0M",
}
`;

exports[`utils / cssSnapshots with file 'test.module.scss' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {
'local-class-inside-global': string;
  'local-class': string;
  'local-class-2': string;
  'local-class-inside-local': string;
  'reserved-words': string;
  'default': string;
  'const': string;
  'nested-class-parent': string;
  'child-class': string;
  'nested-class-parent--extended': string;
  'section-1': string;
  'section-2': string;
  'section-3': string;
  'section-4': string;
  'section-5': string;
  'section-6': string;
  'section-7': string;
  'section-8': string;
  'section-9': string;
  'class-with-mixin': string;
};
export default classes;

export const __cssModule: true;
export type AllClassNames = 'local-class-inside-global' | 'local-class' | 'local-class-2' | 'local-class-inside-local' | 'reserved-words' | 'default' | 'const' | 'nested-class-parent' | 'child-class' | 'nested-class-parent--extended' | 'section-1' | 'section-2' | 'section-3' | 'section-4' | 'section-5' | 'section-6' | 'section-7' | 'section-8' | 'section-9' | 'class-with-mixin';"
`;

exports[`utils / cssSnapshots with file 'test.module.styl' createExports should create an exports file 1`] = `
"declare const classes: {
'foo': string;
  'bar': string;
  'baz': string;
  'col-1': string;
  'col-2': string;
  'col-3': string;
  'local-class-1': string;
  'inside-local': string;
  'inside-1-name-2': string;
  'inside-2-name-1': string;
};
export default classes;
export const foo: string;
export const bar: string;
export const baz: string;
"
`;

exports[`utils / cssSnapshots with file 'test.module.styl' getClasses should return an object matching expected CSS 1`] = `
Object {
  "bar": "test-module__bar---26YdQ",
  "baz": "test-module__baz---3qOSE",
  "col-1": "test-module__col-1---3-pEN",
  "col-2": "test-module__col-2---1JvKR",
  "col-3": "test-module__col-3---2WNwC",
  "foo": "test-module__foo---2CbfH",
  "inside-1-name-2": "test-module__inside-1-name-2---1uX0t",
  "inside-2-name-1": "test-module__inside-2-name-1---1asqx",
  "inside-local": "test-module__inside-local---1XSdE",
  "local-class-1": "test-module__local-class-1---1TLvx",
}
`;

exports[`utils / cssSnapshots with file 'test.module.styl' with a custom template should transform the generated dts 1`] = `
"/* eslint-disable */
declare const classes: {
'foo': string;
  'bar': string;
  'baz': string;
  'col-1': string;
  'col-2': string;
  'col-3': string;
  'local-class-1': string;
  'inside-local': string;
  'inside-1-name-2': string;
  'inside-2-name-1': string;
};
export default classes;
export const foo: string;
export const bar: string;
export const baz: string;

export const __cssModule: true;
export type AllClassNames = 'foo' | 'bar' | 'baz' | 'col-1' | 'col-2' | 'col-3' | 'local-class-1' | 'inside-local' | 'inside-1-name-2' | 'inside-2-name-1';"
`;

exports[`utils / cssSnapshots with includePaths in sass options should find external file from includePaths 1`] = `
Object {
  "big-font": "include-path-module__big-font---3ir6c",
  "class-with-mixin": "include-path-module__class-with-mixin---Ksqo-",
}
`;

exports[`utils / cssSnapshots with includePaths in stylus options should find external file from includePaths 1`] = `
Object {
  "external-class": "include-path-module__external-class---urMTI",
  "include-path": "include-path-module__include-path---3lUJm",
}
`;
